client/authority-discovery/src/worker.rs

```diff
@@ -322,32 +326,22 @@ where
 				.set(addresses.len().try_into().unwrap_or(std::u64::MAX));
 		}
 
-		let mut serialized_addresses = vec![];
-		schema::AuthorityAddresses { addresses }
-			.encode(&mut serialized_addresses)
-			.map_err(Error::EncodingProto)?;
+		let serialized_record = serialize_authority_record(addresses)?;
+		let peer_signature =
+			sign_record_with_peer_id(&serialized_record, &self.network.local_identity())?;
 
 		let keys_vec = keys.iter().cloned().collect::<Vec<_>>();
-		let signatures = key_store
-			.sign_with_all(
-				key_types::AUTHORITY_DISCOVERY,
-				keys_vec.clone(),
-				serialized_addresses.as_slice(),
-			)
-			.await
-			.map_err(|_| Error::Signing)?;
 
-		for (sign_result, key) in signatures.into_iter().zip(keys_vec.iter()) {
-			let mut signed_addresses = vec![];
+		let mut kv_pairs = sign_record_with_authority_ids(
+			serialized_record,
+			Some(peer_signature),
+			key_store.as_ref(),
+			keys_vec,
+		)
+		.await?;
 
-			// Verify that all signatures exist for all provided keys.
-			let signature =
-				sign_result.ok().flatten().ok_or_else(|| Error::MissingSignature(key.clone()))?;
-			schema::SignedAuthorityAddresses { addresses: serialized_addresses.clone(), signature }
-				.encode(&mut signed_addresses)
-				.map_err(Error::EncodingProto)?;
-
-			self.network.put_value(hash_authority_id(key.1.as_ref()), signed_addresses);
+		for (key, value) in kv_pairs.drain(..) {
```

Comment:
```suggestion
		for (key, value) in kv_pairs.into_iter() {
```

Reply:
Both `key` and `value` parameters are owned in `put_value()`, so I rather move them in than `clone` them in and then `drop` them here.

Reply:
What? `into_iter` also gives you the owned types.

Reply:
Oh, not `iter()`, but `into_iter()`. Okay :blush: